### ISSUE TITLE
Stop at the 'Loading json ...'

### ISSUE THREAD

**Issue by callmefish** [Dobiasd/frugally-deep#142]
Hi Tobias, thanks for this great library!
I trained a ResNet50 network using Keras. I was able to convert the .h5 model to a .json. However, when I run the program as follows:
```
#include <fdeep/fdeep.hpp>
#include <opencv2/opencv.hpp>

int main()
{
	const cv::Mat image = cv::imread("Image_1_2.jpg");
	cv::cvtColor(image, image, cv::COLOR_BGR2RGB);
	assert(image.isContinuous());
	const auto model = fdeep::load_model("train7.json");
	// Use the correct scaling, i.e., low and high.
	const auto input = fdeep::tensor5_from_bytes(image.ptr(),
		static_cast<std::size_t>(image.rows),
		static_cast<std::size_t>(image.cols),
		static_cast<std::size_t>(image.channels()),
		0.0f, 1.0f);
	const auto result = model.predict_class({ input });
	std::cout << result << std::endl;
	system("pause");
}
```
It likes the example in the FAQ--How to use images loaded with OpenCV as input for a model?
But it doesn't work with my Keras model. It just spent about 236s to load json, and then stop here. My CPU is Core i5-3230M, which is not a good CPU. My model is used to classify 7 kinds of algae cells, which used transfer learning based on ResNet50.  
The python program for trainning model as follows:
```javascript
import numpy as np
import matplotlib.pyplot as plt
import keras
from keras.preprocessing import image
from keras.preprocessing.image import ImageDataGenerator
from keras.applications import ResNet50
from keras.applications.resnet50 import preprocess_input
from keras import Model, layers
from keras.models import load_model

input_path = "data/LvsRod/"

train_datagen = ImageDataGenerator(
    rescale=1. / 255,
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    preprocessing_function=preprocess_input)

train_generator = train_datagen.flow_from_directory(
    input_path + 'train',
    batch_size=10,
    class_mode='binary',
    target_size=(224, 224))

validation_datagen = ImageDataGenerator(
    rescale=1. / 255,
    preprocessing_function=preprocess_input)

validation_generator = validation_datagen.flow_from_directory(
    input_path + 'validation',
    shuffle=False,
    class_mode='binary',
    target_size=(224, 224))

conv_base = ResNet50(include_top=False, weights='imagenet', input_shape=(224, 224, 3))

for layer in conv_base.layers:
    layer.trainable = False

x = conv_base.output
x = layers.Flatten()(x)
x = layers.Dense(256, activation='relu')(x)
x = layers.Dropout(0.5)(x)
predictions = layers.Dense(7, activation='softmax')(x)
model = Model(conv_base.input, predictions)

optimizer = keras.optimizers.SGD(lr=1e-4, momentum=0.9)
model.compile(loss='sparse_categorical_crossentropy',
              optimizer=optimizer,
              metrics=['accuracy'])

history = model.fit_generator(generator=train_generator,
                              steps_per_epoch=10,  # added in Kaggle
                              epochs=30,
                              validation_data=validation_generator,
                              validation_steps=10  # added in Kaggle
                             )

# save
model.save('train7.h5')
```
The h5model can download from this 

> URL:https://pan.baidu.com/s/1YkuBHBkjjUs2dcpc8XTLqA  
Extraction code：1od1 

Because the file is too big, so I cannot upload it here.  I really want to know how to solve the problem.

**Comment by Dobiasd** [comment]
Sorry, I did not manage to download the h5 file from the page you linked. I randomly clicked some of the buttons with Chinese text on them, and then it started to download some 30 MB `.exe` file, which probably is not what I want. :wink:

> It just spent about 236s to load json

Does it throw some exception or do you see any error?

Anyways, please check if your compiler is optimizing for speed, see here: https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#why-is-my-prediction-roughly-100-times-slower-in-c-as-in-python

If you are using Visual C++, this just means: Use "release" mode instead of "debug" mode.

**Comment by callmefish** [comment]
Well done! Thanks for your patience.  I can't believe that it is so easy to use it. 
![image](https://user-images.githubusercontent.com/31463079/58676647-88c7d080-838b-11e9-8de5-8ec6e5a01614.png)

**Comment by Dobiasd** [comment]
:+1:

**Comment by callmefish** [comment]
Hi, I am coming back for another question. I find the prediction from keras and fdeep are different. I carefully keep the fdeep the same with Keras, but the result is not correct. I hope that you can help me to find the solution. Thx.
the .py for prediction of  pic:
```
import numpy as np
from keras.preprocessing import image
from keras.applications.resnet50 import preprocess_input
from keras.models import load_model

model = load_model('Test2_Resnet.h5')
validation3 = 'Image_1_2.jpg'
img_list0 = image.load_img(validation3)
validation_batch0 = np.stack([preprocess_input(np.array(img_list0))])
pred_probs = model.predict(validation_batch0)
print(np.argmax(pred_probs))
``` 
The result of it as follow:
![image](https://user-images.githubusercontent.com/31463079/58746349-ef361700-848f-11e9-82ec-b997617835ef.png)

The .cpp for prediction of pic；
```
#include <fdeep/fdeep.hpp>
#include <opencv2/opencv.hpp>

int main()
{
	cv::Mat image = cv::imread("Image_1_2.jpg");
	image.convertTo(image, CV_32FC3);
	cv::Mat dst(image.size(), CV_32FC3);
	// preprocess_input in Keras
	for (int r = 0; r<image.rows; r++)
	{
		for (int c = 0; c<image.cols; c++)
		{
			dst.at<cv::Vec3f>(r, c)[0] = image.at<cv::Vec3f>(r, c)[0] - 103.939;
			dst.at<cv::Vec3f>(r, c)[1] = image.at<cv::Vec3f>(r, c)[1] - 116.779;
			dst.at<cv::Vec3f>(r, c)[2] = image.at<cv::Vec3f>(r, c)[2] - 123.68;
		}
	}
	assert(dst.isContinuous());
	const auto model = fdeep::load_model("Test2_Resnet.json");
	// keep[0, 255] float32
	const auto input = fdeep::tensor5_from_bytes(dst.ptr(),
		static_cast<std::size_t>(dst.rows),
		static_cast<std::size_t>(dst.cols),
		static_cast<std::size_t>(dst.channels()),
		0.000f, 255.000f);

	const int result = model.predict_class({ input });
	std::cout << result << std::endl;
	system("pause");
}
```
The result of it as follow；
![image](https://user-images.githubusercontent.com/31463079/58746399-5b187f80-8490-11e9-8a29-12c8907bc186.png)

Thanks again for your fdeep!

**Comment by Dobiasd** [comment]
Please check out this part of the FAQ: https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#why-does-my-model-return-different-values-with-frugally-deep-compared-to-keras

**Comment by Dobiasd** [comment]
Oh, and that part regarding your question about multiple predictions: https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#does-frugally-deep-support-multiple-cpus

**Comment by callmefish** [comment]
Thanks for your reply. The shape of input is different. And I don't know how to check just several elements of the array of input. 
The result of python as follow:
![image](https://user-images.githubusercontent.com/31463079/58748204-9b372c80-84a7-11e9-91f2-e863b5cba181.png)
The result of C++ as follow:
![image](https://user-images.githubusercontent.com/31463079/58748237-29131780-84a8-11e9-9c15-93d2d369d108.png)
Obviously, the element of input in c++ is not correct, because it should not be the zeros behind the decimal point. However, I checked my .cpp and found that I have converted the cv::Mat to CV_32FC3. Maybe it is because the image input for the fdeep::tensor5_from_bytes is limited in uint8_t.
![image](https://user-images.githubusercontent.com/31463079/58748322-75ab2280-84a9-11e9-91c4-c00e545d45d5.png)

**Comment by Dobiasd** [comment]
The shape is the same. It is `(224, 224, 3)` in Python and C++. You can ignore the leading `1`s.

> I don't know how to check just several elements of the array of input.

It's an `fdeep::tensor5`. You can access elements of it by using it's public member function `fdeep::tensor5::get`: https://github.com/Dobiasd/frugally-deep/blob/ceea8ec94521213a99fed07ae280b993260b3912/include/fdeep/tensor5.hpp#L49-L57

**Comment by callmefish** [comment]
Sorry, as a beginner, I can't understand how to use _fdeep::tensor5::get_. Can you give an easy example? Thx.

**Comment by Dobiasd** [comment]
```cpp
#include "fdeep/fdeep.hpp"

int main()
{
    fdeep::tensor5 my_tensor(fdeep::shape5(1, 1, 1, 1, 4), {1, 2, 3, 4});
    float some_value = my_tensor.get(0, 0, 0, 0, 2);
    std::cout << some_value << std::endl;
}
```

**Comment by callmefish** [comment]
Get it! will try it immediately.

**Comment by callmefish** [comment]
Thanks for your reply. I think the problem is the _fdeep::tensor5_from_bytes_. It limits the type of input image, which can only be _uint8-t_ . Therefore, the step of preprocess based on type of _float32_ can not succeed, and the prepocess just brings errors in format conversion. Are there any solution to solve it?

**Comment by callmefish** [comment]
Or is there any way to change the value of tensor5 input? Like change the input(0, 0, 100, 100, 0) from 234 to 112.

**Comment by Dobiasd** [comment]
As its documentation states, `fdeep::tensor5_from_bytes` takes a pointer to the pixel-memory block of an image with 8 bits per pixel (and channel), i.e., unsigned integers of one byte each. The image you are passing to it however if of type `CV_32FC3`, so it has 32 bit (i.e., 4 bytes) for each channel of each pixel.

> Is there any solution to solve it?

For the float operations you want to do, you have two options:

1) Use `tensor5_from_bytes` on the image as it is after loading from disk, probably `CV_8UC3`. Then do the value manipulations on the tensor itself by looping over its entries and use `tensor5::get` and `tensor5::set`.

2) You create a `tensor5` on your own (`const auto my_tensor = fdeep::tensor5 my(fdeep::shape5(1, 1, 224, 224, 3));`) loop over your OpenCV image (already converted to `CV_32FC3`) and use `tensor5::set` to set the manipulated values in the tensor.

---

> Or is there any way to change the value of tensor5 input? Like change the input(0, 0, 100, 100, 0) from 234 to 112.

Yes, `fdeep::tensor5::set`. :)
https://github.com/Dobiasd/frugally-deep/blob/ceea8ec94521213a99fed07ae280b993260b3912/include/fdeep/tensor5.hpp#L78-L86

**Comment by callmefish** [comment]
It works!! So amazing! You created a very useful and powerful tool. It greatly reduces the difficulty of transfering model from Keras. Thank you for your help and timely response. 
Code:
```
#include <fdeep/fdeep.hpp>
#include <opencv2/opencv.hpp>
#include<fstream>

int main()
{
	const auto model = fdeep::load_model("Test2_Resnet.json");
	std::string imgadd = "Image_1_2.jpg";
	cv::Mat image = cv::imread(imgadd);
	const fdeep::tensor5 input = fdeep::tensor5_from_bytes(image.ptr(),
		static_cast<std::size_t>(image.rows),
		static_cast<std::size_t>(image.cols),
		static_cast<std::size_t>(image.channels()),
		0.0f, 255.0f);
	// preprocess_input in Keras, B - 103.939，G - 116.779，R - 123.68
	auto my_tensor = fdeep::tensor5(fdeep::shape5(1, 1, 224, 224, 3), 0.000f);
	for (int r = 0; r < 224; r++)
	{
		for (int c = 0; c < 224; c++)
		{
			my_tensor.set(0, 0, r, c, 0, input.get(0, 0, r, c, 0) - 103.939);
			my_tensor.set(0, 0, r, c, 1, input.get(0, 0, r, c, 1) - 116.779);
			my_tensor.set(0, 0, r, c, 2, input.get(0, 0, r, c, 2) - 123.68);
		}
	}	
	const auto result = model.predict_class({ my_tensor });
	std::ofstream outfile;// create a file
	outfile.open("data.txt");
	outfile << result << std::endl;
	outfile.close();
	std::cout << result << std::endl;
	system("pause");
}
```
Result:
![image](https://user-images.githubusercontent.com/31463079/58756275-e002a800-8527-11e9-998f-d1fdf60e22a5.png)
It is the same as the result of Python.

**Comment by Maksimkhokhovk** [comment]
I have a problem when my code Build json file.
My Resnet model was trained by tensorflow 2.4 and I'm using visual studio 2019(version 16)
![image](https://user-images.githubusercontent.com/86114708/128958178-eed30550-aa73-4b97-87a4-13594e86b332.png)

**Comment by Dobiasd** [comment]
@Maksimkhokhovk

The exception message (not shown in your screenshot) should explain the problem. So I downloaded your model and tried it locally. The result:

```
Loading json ... done. elapsed time: 0.951166 s
Building model ... terminate called after throwing an instance of 'std::runtime_error'
  what():  unknown layer type: RandomContrast
[1]    45684 abort (core dumped)  ./applications_performance
```

Yeah, the quite new [`RandomContrast` layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/RandomContrast) was not yet supported.

But the docs state:

> Contrast is adjusted independently for each channel of each image during training.

So it seems this layer is passive during prediction (like, for example, `Dropout` or `RandomRotation`). So it was easy to add support for it to frugally-deep, which I just did (along with some more of these "no-op" layers). This can be found in [release `v0.15.10-p0`](https://github.com/Dobiasd/frugally-deep/releases/tag/v0.15.10-p0). :slightly_smiling_face:

However, your model also uses a `SlicingOpLambda` layer. This is not supported in frugally-deep. In case it is feasible to remove it from your model, that would be the simplest way for you to make your model compatible with frugally-deep. If you need this functionality, maybe you can emulate it with a [custom layer](https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#how-to-use-custom-layers), but that's pretty advanced.

**Comment by Maksimkhokhovk** [comment]
Thank you very much to reply on me.
what I want is that prediction accuracy using this model should same as before.
Hope your kind help.

**Comment by Dobiasd** [comment]
> what I want is that prediction accuracy using this model should same as before.

Then, I suggest trying the following three possible solutions:
1) Remove `SlicingOpLambda` from your model and check if the accuracy is still fine. If so, you can simply go ahead and use frugally-deep.
2) Try to replace `SlicingOpLambda` in your model with a custom layer that does the same thing, and re-implement it in C++/frugally deep, as hinted [here](https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#how-to-use-custom-layers).
3) Don't use frugally-deep for running the predictions on your model, but something else, maybe TensorFlow lite.

**Comment by Maksimkhokhovk** [comment]
Thanks Dobiasd.
I've read your report, but I'm beginner in cmake project.
I am run my code in Console App project, so I don't know how to make cmake project based on CMakeLists.txt.
I can't execute code you sent me. I don't know how to make success down code.
find_package(FunctionalPlus CONFIG REQUIRED)
![image](https://user-images.githubusercontent.com/86114708/128991349-6d042469-f258-4a09-9431-f62a169f2805.png)

I would like you to descript about this, and if you are complicated with it, you may teach me how to add your cmake your code to my Console App project.
And please teach about second description more deeply.

**Comment by Dobiasd** [comment]
@Maksimkhokhovk Since I did not send you any code, I guess you mean the few lines shown in the FAQ. You were already able to compile/build/run a call to `fdeep::load_model`, so you should already have a VS project set up correctly, which you can re-use for other tests. In general, yeah, CMake sometimes can be tricky. So can be modifying your model and replacing layers while keeping accuracy. I'm sorry, but I can't mentor you on these things in my spare time. If there is a bug in frugally-deep, please continue to let me know. But for all other things, i.e., general programming problems unrelated to frugally-deep, I recommend asking on StackOverflow or trying to find another experienced developer to help you.

**Comment by Maksimkhokhovk** [comment]
Hi Sir, I'm Maksim.
You updated your frugally-deep, and I retry with you updated Lib(I deleted SlicingOpLambda in my model, of course).
But It still can't build my model.
Same error occured.
I wonder if more layers can't be supported consist.
I want your help about this.

**Comment by Dobiasd** [comment]
Since you likely still don't know how to see the exception message, could you please upload your updated model, so I can check?

**Comment by Dobiasd** [comment]
I just checked [your original model](https://sendgb.com/kgfVEjC76d5), and it contains a `TFOpLambda` layer. This one is not supported by frugally-deep.

I highly recommend you find out how to make your IDE show the exception message. Basically, if you compile and run the following

```cpp
#include <stdexcept>
int main()
{
    throw std::runtime_error("hello exception");
}
```

you should be able to see/find `hello exception` somewhere on your screen (not just in the code). That way you could read the exceptions thrown by frugally-deep, which in that case very likely is `unknown layer type: TFOpLambda`.

**Comment by Maksimkhokhovk** [comment]
I tried your simple code but I get exceptions on my code.
It is very strange.
And I tried to test with json model no contain TFOpLambda layer, but occured same error.
![error](https://user-images.githubusercontent.com/86114708/129115112-2be161f3-ec64-41d0-a928-e9057c49f751.png)

Is there any other layers on my model?

**Comment by Dobiasd** [comment]
With you seemingly ignoring my suggestions, it's really hard to help you. 🤷 So please:
- Make sure you updated your local version of frugally deep to [release v0.15.10-p0](https://github.com/Dobiasd/frugally-deep/releases/tag/v0.15.10-p0).
- Upload your new model, i.e., the one without any `SlicingOpLambda` or `TFOpLambda` layers for me to look into. Ideally in `.h5` format.
- Find out how to make your IDE show the actual message of an [unhandled exception](https://gist.github.com/Dobiasd/ce5008e6d479abc91c8100d36332cef7).
- Open a [new issue](https://github.com/Dobiasd/frugally-deep/issues/new) instead of misusing this very old one for your now kinda unrelated problem. This would improve clarity for future readers and also avoid non-involved users receiving superfluous notifications.

**Comment by Dobiasd** [comment]
When loading your model, the error is `dangling layer reference: tf.nn.bias_add`:

```
Loading json ... done. elapsed time: 0.958485 s
Building model ... done. elapsed time: 1.327833 s
Running test 1 of 1 ... terminate called after throwing an instance of 'std::runtime_error'
  what():  dangling layer reference: tf.nn.bias_add
```

See, it would be really helpful if you would try to understand what an exception is and how you can read the message attached to an exception on your own. :wink:

In case you just manually removed stuff from the JSON file, that does not work. You have to actually have a model in Python without the incompatible layers, save it as `.h5` and then convert it to `.json`.

**Please upload the `.h5` version of your model, so I can check what's going on.**

---

PS: You are still spamming the same old issue that is not related to your problem. Again, please create a new one instead. If you continue to ignore my suggestions I'll likely have to stop helping you and instead ban you from posting in this repository.
